### ISSUE TITLE
Some emulators cannot start

### ISSUE THREAD

**Issue by greenEkatherine** [dotnet/xharness#754]
This is a new issue to track two old ones. It mostly affects old emulators with API 21-24 (https://github.com/dotnet/core-eng/issues/13359#issuecomment-922764873) but also seen on new emulators - less frequently (https://github.com/dotnet/xharness/issues/689)

It seems that emulator is hanging, I'm trying to pass different parameters on start to be sure that it has enough space, memory and use hardware properly. If it doesn't help, work around will be to kill VM in bad shape and rerun on the new one.

**Comment by greenEkatherine** [comment]
I've noticed that this issue reproduced during first runs but after some time all machines in my PR's have emulators up and running. It seems that emulators there require more time to starts and want to learn how long it takes.

**Comment by tkapin** [comment]
> It seems that emulator is hanging, I'm trying to pass different parameters on start to be sure that it has enough space, memory and use hardware properly.

@greenEkatherine could you please elaborate on this?

**Comment by premun** [comment]
@greenEkatherine I have some questions as I don't understand the space much:
1. Do we boot the simulator using some commands after the machine starts or do we create a VM with already running simulator and once the VM starts the emulator is ready?
1. What exactly do we do to validate the emulator is running in OSOB? How do we validate in OSOB pipeline that when we make changes to Android queues, the emulators are running?
1. Do you think it makes sense we should add some ValidateHelix step that would run something on the Android queues?

**Comment by greenEkatherine** [comment]
> @greenEkatherine I have some questions as I don't understand the space much:
> 
> 1. Do we boot the simulator using some commands after the machine starts or do we create a VM with already running simulator and once the VM starts the emulator is ready?
> 2. What exactly do we do to validate the emulator is running in OSOB? How do we validate in OSOB pipeline that when we make changes to Android queues, the emulators are running?
> 3. Do you think it makes sense we should add some ValidateHelix step that would run something on the Android queues?

1. We run emulator after the VM starts.
2. In validation script we check that there are two emulators for x86 and x64 (except queue with API 31 where only x64 is availbale) and that their status is not "offline"
3. Do you mean to do it on every VM? because the issue seems to be flaky so our validation usually passed

**Comment by premun** [comment]
Looking at the code - we seem to be doing some _"waiting for emulators to start"_ in the validation phase. However, we're not doing the same when actually launching the simulators (first-run or when starting the systemd service) since validation scripts don't run then. Do you know if there's some difference or reason for that?

**Comment by akoeplinger** [comment]
FWIW `-use-system-libs` is kinda scary, we shouldn't be using it if the issue is just that the emulator starts more slowly. We don't care if it takes 10mins for a VM+emulator to boot as long as it works :)

I do know that we had issues with the "wait for emulator" code in the past, I wonder if just waiting some hardcoded time on these problematic API levels would be enough for now?

**Comment by premun** [comment]
> I do know that we had issues with the "wait for emulator" code in the past

From the emulator artifact's code it seems that we're not waiting for emulator at all when booting the VM. We only wait for it in validation phase that happens when we're building the image itself (and that's where we've seen the flaky behaviour).

So it might be we just start accepting helix jobs too early.

> I do know that we had issues with the "wait for emulator" code in the past, I wonder if just waiting some hardcoded time on these problematic older API levels would be enough for now?

I guess a plan B? :)

**Comment by akoeplinger** [comment]
I think we thought that it'd be handled by xharness' wait-for-device code, but the 5min timeout for that is probably too short: https://github.com/dotnet/xharness/blob/63e517161fd1ca3855c277edd92f0fafe51ca33b/src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs#L158

**Comment by greenEkatherine** [comment]
I run 100 jobs and set up timeout up to 30 minutes. Unfortunately there are only 4 machines took jobs out of 23 machines  - too few to estimate timeout, but even there I see that it takes 10-15 minutes to run the first jobs on each machine.

I would like to look at the results from bigger number of machines. @premun does `WorkItemWaitTime` monitoring show it or it's about Helix Queueing?

**Comment by premun** [comment]
Afaik `WorkItemWaitTime` is `Started - Enqueued` - so how long it took to pull the work item from the queue after it was enqueued so it is probably not useful here.

You could technically scale up the queue manually in Azure to get more machines and more data - if you don't let them run needlessly long.
But not sure how exact we want to be.

**Comment by greenEkatherine** [comment]
Second attempt show that 19 out of 30 machines take 10+ minutes to run emulators for their first job.

I raised PR and will try to run tests against this queue https://dev.azure.com/dnceng/internal/_git/dotnet-helix-machines/pullrequest/19540

**Comment by premun** [comment]
@greenEkatherine @fanyang-mono reported a case where we see no emulator on the default API (`ubuntu.1804.amd64.android.open.svc`).

All of the work items processed on the machine in the span of almost 15 minutes have failed on the `ADB_ENUMERATION_FAILURE`. However, the last work item made several attempts until almost 30 minutes from the first one.

**Comment by greenEkatherine** [comment]
Current state in PR:
Based on the feedback, I moved waiting for device command with 15 minutes timeout into systemd, so it will be called not only during the first run but after reboot as well. I put it into `ExecStartPost`. If I understand correctly documentation, it will run right after starting emulator but we wait for result before calling `helix.service`.

I would not comment here if it works - it does not. I send 200 test jobs to check emulators' state and more than a half of them failed. The question is why validation always passed and why service is not indicating a failure?
![image](https://user-images.githubusercontent.com/19972657/141444763-2addf75e-e57c-419d-96f0-179a251d2d8c.png)

**Comment by premun** [comment]
I was randomly trying to run XHarness locally and saw this:
![image](https://user-images.githubusercontent.com/7013027/146384016-41dc811b-6d94-4bf1-b10c-4fab667918d8.png)

It might be possible that the emulators actually start fine and XHarness doesn't detect this well?

**Comment by premun** [comment]
Oh, the emulator was actually frozen..

**Comment by missymessa** [comment]
@premun looks like there's still an open PR for this work. Did we want to do anything with it? https://dev.azure.com/dnceng/internal/_git/dotnet-helix-machines/pullrequest/19540

**Comment by premun** [comment]
@missymessa yes, this is very much on my radar but I didn't have time to have a look at this yet. It is very important and Katya has done quite a lot of investigations in this regard. Some of them are in the PR, so I am keeping it until I have time to have a look. I expect to get to this in about 2 weeks.

**Comment by premun** [comment]
I found more proof that the device can reboot on its own in between `android run` commands:
```
Analyzing android/run@x86_64 (0)
2022-04-13T02:45:08.815Z	INFO   	xharness-event-processor.py	event(51)	send	Sending event type _MobileDeviceOperation
2022-04-13T02:45:08.900Z	INFO   	xharness-event-processor.py	saferequests(87)	request_with_retry	Response complete with status code '201'
Analyzing android/run@None (81)
    Encountered DEVICE_NOT_FOUND. This is typically not a failure of the work item. It will be run again. This machine will reboot to heal.
    If this occurs repeatedly, please check for architectural mismatch, e.g. sending arm64_v8a APKs to an x86_64 / x86 only queue.
2022-04-13T02:45:08.901Z	INFO   	xharness-event-processor.py	event(51)	send	Sending event type _MobileDeviceOperation
2022-04-13T02:45:08.936Z	INFO   	xharness-event-processor.py	saferequests(87)	request_with_retry	Response complete with status code '201'
```

One second it's there, XHarness starts it fine, the next second it's gone and it's not XHarness causing the reboot.

This query shows all first operations that started failing:
```sql
Metrics
| where EventType == "_MobileDeviceOperation" and MetricName == "ExitCode" and Timestamp > now()-3d
| extend Dimensions = parse_json(Dimensions)
| extend
    ExitCode = MetricValue,
    Command = tostring(Dimensions.command),
    Platform = tostring(Dimensions.platform),
    Target = tostring(Dimensions.target)
| where Platform == "android"
| join kind=inner WorkItems on WorkItemId
| where MachineName != ""
| order by JobId asc, WorkItemId asc, Timestamp asc
| extend PrevExitCode = prev(ExitCode, 1), PrevCommand = prev(Command, 1)
| where PrevExitCode == 0 and ExitCode == 81
| project
    JobId,
    WorkItemFriendlyName,
    MachineName,
    Target,
    Command,
    PrevCommand,
    ConsoleUri
```

**Comment by tkapin** [comment]
Thanks for the detailed analysis, @premun. Having proved the emulators are actually fixed by rebooting is a good news, also the sudden crash by any operation is a strong clue.

**Comment by premun** [comment]
Furthermore, I added some options in case the emulator start-up would get "stalled" (Android term):
https://dev.azure.com/dnceng/internal/_git/dotnet-helix-machines/pullrequest/22296

I will dig a bit more into the exact moment the emulator stops running before I close this.

**Comment by premun** [comment]
Last piece of investigation was the overall SLA for the emulators. I have only found 4 work items in the last week that failed with on third attempts and none of them for infra reasons.

From the above reasons I conclude that the Android emulators are booting and rebooting fine and I am closing this issue.
